pkg/crosscluster/logical/replication_statements.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package logical
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree/treecmp"
+)
+
+// getPhysicalColumns returns the list of columns that are part of the table's
+// primary key and value.
+func getPhysicalColumns(table catalog.TableDescriptor) []catalog.Column {
+	columns := table.AllColumns()
+	result := make([]catalog.Column, 0, len(columns))
+	for _, col := range columns {
+		if !col.IsComputed() && !col.IsVirtual() && !col.IsSystemColumn() {
+			result = append(result, col)
+		}
+	}
+	return result
+}
+
+// newInsertStatement returns a statement that can be used to insert a row into
+// the table.
+//
+// The statement will have `n` parameters, where `n` is the number of columns
+// in the table. Parameters are ordered by column ID.
+func newInsertStatement(table catalog.TableDescriptor) (tree.Statement, error) {
+	columns := getPhysicalColumns(table)
+
+	columnNames := make(tree.NameList, len(columns))
+	for i, col := range columns {
+		columnNames[i] = tree.Name(col.GetName())
+	}
+
+	parameters := make(tree.Exprs, len(columnNames))
+	for i := range columnNames {
+		var err error
+		parameters[i], err = tree.NewPlaceholder(fmt.Sprintf("%d", i+1))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	parameterValues := &tree.ValuesClause{
+		Rows: []tree.Exprs{
+			parameters,
+		},
+	}
+
+	rows := &tree.Select{
+		Select: parameterValues,
+	}
+
+	insert := &tree.Insert{
+		Table: &tree.TableRef{
+			TableID: int64(table.GetID()),
+			As:      tree.AliasClause{Alias: "replication_target"},
+		},
+		Rows:      rows,
+		Columns:   columnNames,
+		Returning: tree.AbsentReturningClause,
+	}
+
+	return insert, nil
+}
+
+// newMatchesLastRow creates a WHERE clause for matching all columns of a row.
+// It returns a tree.Expr that compares each column to a placeholder parameter.
+// Parameters are ordered by column ID, starting from startParamIdx.
+func newMatchesLastRow(columns []catalog.Column, startParamIdx int) (tree.Expr, error) {
+	var whereClause tree.Expr
+	for i, col := range columns {
+		placeholder, err := tree.NewPlaceholder(fmt.Sprintf("%d", startParamIdx+i))
+		if err != nil {
+			return nil, err
+		}
+		colExpr := &tree.ComparisonExpr{
+			Operator: treecmp.MakeComparisonOperator(treecmp.EQ),
```

Comment:
Here's the one thing I noticed: for nullable columns, this will need to be `treecmp.IsNotDistinctFrom` rather than `treecmp.EQ`.

Reply:
Good catch! I fixed this and updated a WIP test for the PR that uses this change to include a null column.